Sources/_StringProcessing/Regex/DSLTree.swift

```diff
@@ -455,3 +452,225 @@ public struct CaptureTransform: Hashable, CustomStringConvertible {
+  /// `_TreeNode` conformance.
+  struct _Tree: _TreeNode {
+    var node: DSLTree.Node
+    
+    init(_ node: DSLTree.Node) {
+      self.node = node
+    }
+    
+    var children: [_Tree]? {
+      switch node {
+        
+      case let .orderedChoice(v): return v.map(_Tree.init)
+      case let .concatenation(v): return v.map(_Tree.init)
+
+      case let .convertedRegexLiteral(n, _):
+        // Treat this transparently
+        return _Tree(n).children
+
+      case let .capture(_, _, n):           return [_Tree(n)]
+      case let .nonCapturingGroup(_, n):    return [_Tree(n)]
+      case let .transform(_, n):            return [_Tree(n)]
+      case let .quantification(_, _, n):    return [_Tree(n)]
+
+      case let .conditional(_, t, f): return [_Tree(t), _Tree(f)]
+
+      case .trivia, .empty, .quotedLiteral, .regexLiteral,
+          .consumer, .matcher, .characterPredicate,
+          .customCharacterClass, .atom:
+        return []
+
+      case let .absentFunction(abs):
+        return abs.ast.children.map(\.dslTreeNode).map(_Tree.init)
+      }
+    }
+    
+    func _captureStructure(
+      _ constructor: inout CaptureStructure.Constructor
+    ) -> CaptureStructure {
+      switch node {
+      case let .orderedChoice(children):
+        return constructor.alternating(children.map(_Tree.init))
+
+      case let .concatenation(children):
+        return constructor.concatenating(children.map(_Tree.init))
+
+      case let .capture(name, _, child):
+        if let type = child.valueCaptureType {
+          return constructor.capturing(
+            name: name, _Tree(child), withType: type)
+        }
+        return constructor.capturing(name: name, _Tree(child))
+
+      case let .nonCapturingGroup(kind, child):
+        assert(!kind.ast.isCapturing)
+        return constructor.grouping(_Tree(child), as: kind.ast)
+
+      case let .conditional(cond, trueBranch, falseBranch):
+        return constructor.condition(
+          cond.ast,
+          trueBranch: _Tree(trueBranch),
+          falseBranch: _Tree(falseBranch))
+
+      case let .quantification(amount, _, child):
+        return constructor.quantifying(
+          Self(child), amount: amount.ast)
+
+      case let .regexLiteral(re):
+        // TODO: Force a re-nesting?
+        return re.ast._captureStructure(&constructor)
+
+      case let .absentFunction(abs):
+        return constructor.absent(abs.ast.kind)
+
+      case let .convertedRegexLiteral(n, _):
+        // TODO: Switch nesting strategy?
+        return Self(n)._captureStructure(&constructor)
+
+      case .matcher:
+        return .empty
+
+      case .transform(_, let child):
+        return Self(child)._captureStructure(&constructor)
+
+      case .customCharacterClass, .atom, .trivia, .empty,
+          .quotedLiteral, .consumer, .characterPredicate:
+        return .empty
+      }
+    }
+  }
+
+  @_spi(RegexBuilder)
+  public enum _AST {
```

Comment:
Why is the AST relevant at all to DSLTree? Is it because we don't have or want to have versions of these enums?

Reply:
These are the things that we hadn't duplicated into the DSLTree layer, so we were just storing AST symbols in DSLTree nodes. My goal with this change was just to wrap those, not necessarily replace them wholesale. At some point it would be great to have better separation between the two modules.